plugins/inputs/execd/execd.go

```diff
@@ -90,59 +97,62 @@ func (e *Execd) Stop() {
 	e.wg.Wait()
 }
 
-func (e *Execd) cmdLoop(ctx context.Context) {
+// cmdLoop watches an already running process, restarting it when appropriate.
+func (e *Execd) cmdLoop(ctx context.Context) error {
 	for {
 		// Use a buffered channel to ensure goroutine below can exit
 		// if `ctx.Done` is selected and nothing reads on `done` anymore
 		done := make(chan error, 1)
 		go func() {
-			done <- e.cmdRun()
+			done <- e.cmdWait()
 		}()
 
 		select {
 		case <-ctx.Done():
-			e.stdin.Close()
-			// Immediately exit process but with a graceful shutdown
-			// period before killing
-			internal.WaitTimeout(e.cmd, 200*time.Millisecond)
-			return
+			if e.stdin != nil {
+				e.stdin.Close()
+				// Immediately exit process but with a graceful shutdown
+				// period before killing
+				internal.WaitTimeout(e.cmd, 200*time.Millisecond)
+			}
+			return nil
 		case err := <-done:
 			log.Printf("E! [inputs.execd] Process %s terminated: %s", e.Command, err)
+			return err
 		}
 
-		log.Printf("E! [inputs.execd] Restarting in %s...", e.RestartDelay.Duration)
+		log.Printf("E! [inputs.execd] Restarting in %s...", time.Duration(e.RestartDelay))
 
 		select {
 		case <-ctx.Done():
-			return
-		case <-time.After(e.RestartDelay.Duration):
+			return nil
+		case <-time.After(time.Duration(e.RestartDelay)):
 			// Continue the loop and restart the process
+			if err := e.cmdStart(); err != nil {
+				return err
```

Comment:
If we return here the process will never be restarted, I think we need to log, wait another RestartDelay, and try to start it again.

Reply:
So then... what do we care if start returns the start error or not? Maybe I should go back to how I had this in the cmdLoop

Reply:
Anytime the subprocess exits the execd input should restarts it, we want to retry forever just like we would if an output plugin was unable to connect.  Exception being the error from Exec on startup and that is to catch and report configure time errors.

Reply:
if it fails on this line it's probably a configuration error and it'll never start. I guess the question is do you want it to keep spamming the logs with retries to what is likely a config error?